### PR TITLE
feat: add support for Apple Silicon chromium builds

### PR DESCRIPTION
**Pull request by 7rulnik**
Google has published [Chromium builds for Apple Silicon](https://commondatastorage.googleapis.com/chromium-browser-snapshots/index.html?prefix=Mac_Arm/) so we can fetch it now

Related to https://github.com/puppeteer/puppeteer/issues/6622

### CODE REVIEW

**Comment by 7rulnik** [comment]
@googlebot I signed it!

**Comment by 7rulnik** [comment]
I got 3 tests failed, but not sure that it's related to this PR

```
  1) Accessibility
       roledescription:
     Error: expect(received).toEqual(expected) // deep equality

Expected: "foo"
Received: undefined
      at Context.<anonymous> (test/accessibility.spec.ts:171:50)
      at processTicksAndRejections (internal/process/task_queues.js:95:5)

  2) network
       Response.fromCache
         should work:
     Error: expect(received).toBe(expected) // Object.is equality

Expected: 304
Received: 200
      at Context.<anonymous> (test/network.spec.ts:166:56)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (internal/process/task_queues.js:95:5)

  3) network
       Page.authenticate
         should not disable caching:
     Error: expect(received).toBe(expected) // Object.is equality

Expected: 304
Received: 200
      at Context.<anonymous> (test/network.spec.ts:606:56)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (internal/process/task_queues.js:95:5)
```

File: src/node/BrowserFetcher.ts
Comment (7rulnik) [inline src/node/BrowserFetcher.ts]: it doesn't feel right for me, but firefox ships universal build, so in `downloadURLs` we can't use the separate url for firefox for MacOS ARM because of string interpolation

Comment (7rulnik) [reply src/node/BrowserFetcher.ts]: as workaround we can pass platform as query parameter in firefox case and remove this hack

**Comment by jschfflr** [comment]
I can confirm that the three test failures you described are related to Chromium changes that we haven't rolled into Puppeteer yet.

**Comment by jschfflr** [comment]
The change looks good to me from a technical standpoint but we'll have to wait for the next Chromium roll to land it.
That one is currently blocked on some network related changes but I'll make sure to include this change with the next roll.
Thanks so much for your contribution!

**Comment by 7rulnik** [comment]
Sure! Feel free to ping me. Thanks

**Comment by maksimr** [comment]
Any chance to merge this PR?

Related issue - https://github.com/puppeteer/puppeteer/issues/7632

**Comment by joshbranham** [comment]
Would love to see this land. As an alternative approach, would anyone be opposed to this [function](/opt/homebrew/bin/chromium) searching `PATH` for chromium? In our case `brew install chromium` drops it in `/opt/homebrew/bin/chromium` per Homebrew on Apple Silicon.

**Comment by SpawnAtis** [comment]
any progress ?

**Comment by 7rulnik** [comment]
@jschfflr as I understand, we are not blocked anymore. Should I update PR?

https://commondatastorage.googleapis.com/chromium-browser-snapshots/index.html?prefix=Mac/938248/
https://commondatastorage.googleapis.com/chromium-browser-snapshots/index.html?prefix=Mac_Arm/938248/

**Comment by 7rulnik** [comment]
Also, as I understand, it would be nice to test against MacOS ARM? But it's currently blocked by https://github.com/actions/virtual-environments/issues/2187

**Comment by OrKoN** [comment]
Let's wait for GitHub Actions to [support it](https://github.com/actions/runner/issues/805) or we can we land it as an experimental opt-in feature?

**Comment by 7rulnik** [comment]
@OrKoN honestly I prefer to release it as an opt-in because we don't really have any timeline for GitHub Actions release. Maybe we can download the ARM version only if env variables set? Something like `PUPPETEER_CHROMIUM_MAC_ARM`?

**Comment by OrKoN** [comment]
yes, I think an env var would work. Perhaps `PUPPETEER_EXPERIMENTAL_CHROMIUM_MAC_ARM`?

**Comment by 7rulnik** [comment]
I will update PR on this weekend or next week

**Review by OrKoN — CHANGES_REQUESTED** [review]
marking as requiring changes. Thanks for working on this!

**Comment by 7rulnik** [comment]
@OrKoN does it look good?

Also I checked that the current chrome version exists https://commondatastorage.googleapis.com/chromium-browser-snapshots/index.html?prefix=Mac_Arm/961656/

**Comment by OrKoN** [comment]
Some errors seem to be related to the Firefox binaries.

**Comment by artyil** [comment]
Is there any updates?

**Comment by 7rulnik** [comment]
Will try to finish it on next week

**Comment by artyil** [comment]
Hi. Can I test if this change actually solve my problem? Until it merged… Is there some kind of pre-release version i can install in package.json to test it?

**Comment by 7rulnik** [comment]
@OrKoN could you trigger actions please?

**Comment by 7rulnik** [comment]
I ran tests locally and this is what I got:

<details>
<summary>11 tests failed within `npm run unit`</summary>

```
  1) Emulation
       Page.emulateCPUThrottling
         should change the CPU throttling rate successfully:
     Error: Timeout of 25000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/Users/v7rulnik/projects/oss/puppeteer/test/emulation.spec.ts)
      at listOnTimeout (internal/timers.js:557:17)
      at processTimers (internal/timers.js:500:7)

  2) headful tests
       HEADFUL
         background_page target type should be available:
     Error: Failed to launch the browser process!
assertion failed [precise_state_mode == PreciseStateMode::UpdateLiveState]: not updating live register state but no fpr state passed in
(ThreadContextRegisterState.cpp:559 move_to_instruction_boundary)


TROUBLESHOOTING: https://github.com/puppeteer/puppeteer/blob/main/docs/troubleshooting.md

      at onClose (src/node/BrowserRunner.ts:268:9)
      at ChildProcess.<anonymous> (src/node/BrowserRunner.ts:255:61)
      at ChildProcess.emit (domain.js:470:12)
      at Process.ChildProcess._handle.onexit (internal/child_process.js:277:12)

  3) Mouse
       should send mouse wheel events:
     Error: expect(received).toMatchObject(expected)

- Expected  - 2
+ Received  + 2

  Object {
-   "height": 230,
-   "width": 230,
+   "height": 345,
+   "width": 345,
  }
      at Context.<anonymous> (test/mouse.spec.ts:198:30)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (internal/process/task_queues.js:95:5)

  4) Screenshots
       ElementHandle.screenshot
         should take into account padding and border:
     Error: screenshot-element-padding-border.png mismatch! Output is saved in "output-chromium" directory
      at Context.<anonymous> (test/screenshot.spec.ts:215:26)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (internal/process/task_queues.js:95:5)

  5) Screenshots
       ElementHandle.screenshot
         should capture full element when larger than viewport:
     Error: screenshot-element-larger-than-viewport.png mismatch! Output is saved in "output-chromium" directory
      at Context.<anonymous> (test/screenshot.spec.ts:241:28)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (internal/process/task_queues.js:95:5)

  6) Screenshots
       ElementHandle.screenshot
         should scroll element into view:
     Error: screenshot-element-scrolled-into-view.png mismatch! Output is saved in "output-chromium" directory
      at Context.<anonymous> (test/screenshot.spec.ts:276:26)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (internal/process/task_queues.js:95:5)

  7) Screenshots
       ElementHandle.screenshot
         should work for an element with fractional dimensions:
     Error: screenshot-element-fractional.png mismatch! Output is saved in "output-chromium" directory
      at Context.<anonymous> (test/screenshot.spec.ts:327:26)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (internal/process/task_queues.js:95:5)

  8) Screenshots
       ElementHandle.screenshot
         should work for an element with an offset:
     Error: screenshot-element-fractional-offset.png mismatch! Output is saved in "output-chromium" directory
      at Context.<anonymous> (test/screenshot.spec.ts:337:26)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (internal/process/task_queues.js:95:5)

  9) Tracing
       should output a trace:
     Error: Timeout of 25000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/Users/v7rulnik/projects/oss/puppeteer/test/tracing.spec.ts)
      at listOnTimeout (internal/timers.js:557:17)
      at processTimers (internal/timers.js:500:7)

  10) Tracing
       should work without options:
     Error: Timeout of 25000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/Users/v7rulnik/projects/oss/puppeteer/test/tracing.spec.ts)
      at listOnTimeout (internal/timers.js:557:17)
      at processTimers (internal/timers.js:500:7)

  11) Tracing
       should support a buffer without a path:
     Error: Timeout of 25000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/Users/v7rulnik/projects/oss/puppeteer/test/tracing.spec.ts)
      at listOnTimeout (internal/timers.js:557:17)
      at processTimers (internal/timers.js:500:7)
```
</details>

----------

<details>
<summary>6 tests failed within `PUPPETEER_EXPERIMENTAL_CHROMIUM_MAC_ARM=1 npm run unit`</summary>

```
  1) Mouse
       should send mouse wheel events:
     Error: expect(received).toMatchObject(expected)

- Expected  - 2
+ Received  + 2

  Object {
-   "height": 230,
-   "width": 230,
+   "height": 345,
+   "width": 345,
  }
      at Context.<anonymous> (test/mouse.spec.ts:198:30)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (internal/process/task_queues.js:95:5)

  2) Screenshots
       ElementHandle.screenshot
         should take into account padding and border:
     Error: screenshot-element-padding-border.png mismatch! Output is saved in "output-chromium" directory
      at Context.<anonymous> (test/screenshot.spec.ts:215:26)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (internal/process/task_queues.js:95:5)

  3) Screenshots
       ElementHandle.screenshot
         should capture full element when larger than viewport:
     Error: screenshot-element-larger-than-viewport.png mismatch! Output is saved in "output-chromium" directory
      at Context.<anonymous> (test/screenshot.spec.ts:241:28)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (internal/process/task_queues.js:95:5)

  4) Screenshots
       ElementHandle.screenshot
         should scroll element into view:
     Error: screenshot-element-scrolled-into-view.png mismatch! Output is saved in "output-chromium" directory
      at Context.<anonymous> (test/screenshot.spec.ts:276:26)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (internal/process/task_queues.js:95:5)

  5) Screenshots
       ElementHandle.screenshot
         should work for an element with fractional dimensions:
     Error: screenshot-element-fractional.png mismatch! Output is saved in "output-chromium" directory
      at Context.<anonymous> (test/screenshot.spec.ts:327:26)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (internal/process/task_queues.js:95:5)

  6) Screenshots
       ElementHandle.screenshot
         should work for an element with an offset:
     Error: screenshot-element-fractional-offset.png mismatch! Output is saved in "output-chromium" directory
      at Context.<anonymous> (test/screenshot.spec.ts:337:26)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (internal/process/task_queues.js:95:5)
```

</details>

BTW should we store ARM screenshots in a different folder?

---------

All tests passed within `PUPPETEER_PRODUCT=firefox npm run unit`

----------

So for me, all looks good, and failed tests are not related to this PR. I ran tests in the main branch and got almost the same results.

**Comment by 7rulnik** [comment]
@OrKoN I think I found a bug. Could you trigger one more time, please?

File: src/node/install.ts
Comment (OrKoN) [inline src/node/install.ts:105]: I think you just need to rebase and remove the line with the os import above. It appears that all tests would pass then.

Comment (7rulnik) [reply src/node/install.ts:105]: fixed it and solved conflicts

**Comment by OrKoN** [comment]
Looks good, thanks for the contribution! May I ask you to update the documentation please? E.g., I could think of mentioning the new flag as experimental here https://github.com/puppeteer/puppeteer/blob/main/docs/api.md#environment-variables

**Comment by 7rulnik** [comment]
@OrKoN added some info to `docs/api.md` and to `contributing.md` Feel free to change wordings if they are not correct

**Comment by cdeutsch** [comment]
Great to see this close to landing.

I'm guessing this won't help people running puppeteer in Docker on Apple Silicon? 
https://github.com/garris/BackstopJS/issues/1300

Are there Arm builds of Chromium for Debian running in Docker?

**Review by OrKoN — APPROVED** [review]

**Comment by OrKoN** [comment]
@cdeutsch I am not aware of Arm builds for Debian that are available for download (but I might just not know)

@7rulnik thanks for the PR, it looks good  to me!

**Review by OrKoN — APPROVED** [review]

**Comment by artyil** [comment]
Was this one supposed to fix https://github.com/puppeteer/puppeteer/issues/7916 ?

**Comment by 7rulnik** [comment]
@artyil nope. It's not related to any Linux (inside docker too) at all. Just native M1.

**Comment by artyil** [comment]
@7rulnik the referred issue is related to m1 only - “[Bug]: Docker M1 Mac fails to open new page (Target.createTarget) #7916” 
Can you elaborate what am I missing there? You mean this fix completely unrelated to `Docker` issues, only when running on M1 (eg: executing node.js script in terminal)  🙄 ?

**Comment by 7rulnik** [comment]
@artyil this PR is related only to M1. This https://github.com/puppeteer/puppeteer/pull/7546 issue is about docker so it's about chromium and puppeteer on Linux. Also I see this comment https://github.com/puppeteer/puppeteer/issues/7916#issuecomment-1060034559 and looks like in this case it's about native run (without docker) too, but I'm not sure. WIll try to run example from it.

**Pull request merged**
